### ISSUE TITLE
DisplayList does not optimize out unnecessary save/restore pairs

### ISSUE THREAD

**Issue by flar** [flutter/flutter#107579]
Frequently developers will do a protective save/restore around some external code "in case" it modifies the clip/transform, but if that external code doesn't execute any of those calls, the save/restore is not useful and just wastes time.

SkPicture used to optimize out unnecessary save/restore pairs, but DisplayList does not have that optimization.

**Comment by JsouLiang** [comment]
First we synchronize the test results:
Scenario one：
```C++
  DisplayListBuilder builder;
  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
  builder.saveLayer(nullptr, true);           // saved   
  builder.saveLayer(nullptr, true);           // unnecessary
  builder.drawRect({10, 10, 20, 20});
  builder.saveLayer(nullptr, true);           // unnecessary
  builder.drawRect({15, 15, 25, 25});
  builder.restore();
  builder.restore();
  builder.restore();

  builder.Build()->Dispatch(expector);
  EXPECT_EQ(expector.save_layer_count(), 1);
```

```C++

  DisplayListBuilder builder;
  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
  builder.saveLayer(nullptr, true);         // saved
  builder.clipRect(rect);                         // clip or transform...
  builder.saveLayer(nullptr, true);         // saved
  builder.drawRect({10, 10, 20, 20});
  builder.saveLayer(nullptr, true);         // unnecessary
  builder.drawRect({15, 15, 25, 25});
  builder.restore();
  builder.restore();
  builder.restore();

  builder.Build()->Dispatch(expector);
  EXPECT_EQ(expector.save_layer_count(), 2);
```

```C++

  DisplayListBuilder builder;
  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
  builder.saveLayer(nullptr, true);              // saved
  builder.saveLayer(nullptr, true);              // unnecessary
  builder.clipRect(rect);                              // clip or transform...
  builder.drawRect({10, 10, 20, 20});
  builder.saveLayer(nullptr, true);              // saved
  builder.drawRect({15, 15, 25, 25});
  builder.restore();
  builder.restore();
  builder.restore();

  builder.Build()->Dispatch(expector);
  EXPECT_EQ(expector.save_layer_count(), 2);
```

```C++

  DisplayListBuilder builder;
  builder.setColor(SkColorSetARGB(127, 255, 255, 255));
  builder.saveLayer(nullptr, true);                       // saved

  builder.clipRect(rect);                                       // clip or transform...
  builder.drawRect({10, 10, 20, 20});
  builder.saveLayer(nullptr, true);                      // saved

  builder.clipRect(rect);                                      // clip or transform...
  builder.drawRect({10, 10, 20, 20});
  builder.saveLayer(nullptr, true);                       // saved

  builder.clipRect(rect);  // clip or transform...
  builder.drawRect({15, 15, 25, 25});
  builder.restore();
  builder.restore();
  builder.restore();

  builder.Build()->Dispatch(expector);
  EXPECT_EQ(expector.save_layer_count(), 3);
```

**Comment by flar** [comment]
A couple of nits about the DisplayListBuilder code written in the previous comment:

- builder.setColor has a DlColor argument. While I designed DlColor to be cast-compatible with any value that developers used to use to pass around SkColor values, primarily so that the unit tests would not need to all be modified unless we want to spend that time, we should probably write example code to actually use the DlColor mechanism moving forward.
- builder.setAttribute() methods are stateful. It gets really hard to use that stateful mechanism over time as you have to look for the history of attribute setting methods in trying to understand each line of code. As such, it is hard to parse the above sequences and recognize what attributes are in effect for each operation. I added builder.op(..., paint) methods to the Builder class to avoid that attribute state confusion. I'm thinking of deleting all of the old stateful methods at some point in the future so that the stateless interface is the only way to use Builder. It couldn't hurt to use only stateless method calls in pseudo-code for clarity.

**Comment by flar** [comment]
Some of the examples that use a clipRect are hard to gauge because it might matter if the clips are intersecting/containing or not. A new saveLayer will perhaps not be needed if the clip hasn't been reduced since the previous saveLayer.

**Comment by flar** [comment]
Also, why the focus on saveLayers? While it will be great to get rid of unnecessary saveLayer calls there will be a huge number of rules to consider for that and the much more common (though less impactful) case is unnecessary save/restore pairs. That was the primary aim of this Issue when I wrote it and a much simpler set of rules to govern their elimination.

**Comment by JsouLiang** [comment]
I have an idea.
We can do something about it in the saveLayer operation of display_list_builder
In the saveLayer operation, we determine if we need to `push<SaveLayerOpt>` based on the `need_save_layer` property of layer_info_.
1. if the current layer is set to transform, `need_save_layer = true`
2. if clip is set and the rectangle of clip intersects with the bounds of current_layer, `need_save_layer = true`

**Comment by flar** [comment]
We can also do a simplifying pass over all of the records before we finalize it, so marking them like we do with SaveLayerOptions can help us identify which are not necessary and then we can do a last reduction pass in Build().

Identifying unnecessary layers is much more complicated than I think you identified. For example, blend mode matters because blending into randomly colored pixels from the outer layer may have a different outcome than blending into the transparent pixels of the save layer and then blending the layer pixels into the underlying pixels. The blend mode for each rendering operation and the blend mode for the saveLayer itself matter in that case.

Overlapping primitives in the content vs non-overlapping matters as they affect how alpha and colorfilter might behave.

Those are just off the top of my head. If I started such a project I would have to go through every aspect of the saveLayer process and what items can be in the content with a fine toothed comb to figure out what the restrictions are on determining if a saveLayer can be optimized out.

There are definitely some conditions that require you to watch the content stream inside the saveLayer to know if you can do an optimization.

I can think of one situation where you can optimize out the whole layer which is alpha=0 and blend=SrcOver (and I'd have to think if other attributes could affect this). In that case then we can drop the layer plus all of its content. But for the general case you need to look at the bigger picture of the entire layer after it is done being built.

**Comment by JsouLiang** [comment]
https://github.com/flutter/engine/pull/34967
